### PR TITLE
feat(snaps): Add support for custom network per Snap

### PR DESCRIPTION
**Pull request by Mrtenz**
## **Description**

Like with websites, we want to support a custom network per Snap rather than relying on the globally selected network. For Snaps, the permission for the network to switch to is automatically granted.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/26389?quickstart=1)

## **Manual testing steps**

1. Copy the build files for `@metamask/selected-network-controller` from [this PR](https://github.com/MetaMask/core/pull/4602).
2. Copy the build files for `@metamask/snaps-controllers` and `@metamask/snaps-execution-environments` from [this PR](https://github.com/MetaMask/snaps/pull/2634).
3. Run `test-snaps` for the PR above (`yarn start` in `packages/test-snaps`).
4. Install the Ethereum Provider Example Snap.
5. Switch the chain ID using the `test-snaps` button for the Ethereum Provider example.
6. The Snap should now use the new network, without any permission requests.

Step 1 and 2 are no longer necessary once these PRs are merged and released.

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: builds.yml [hunk 1]
```diff
@@ -26,7 +26,7 @@ buildTypes:
       - SEGMENT_WRITE_KEY_REF: SEGMENT_PROD_WRITE_KEY
       - ALLOW_LOCAL_SNAPS: false
       - REQUIRE_SNAPS_ALLOWLIST: true
-      - IFRAME_EXECUTION_ENVIRONMENT_URL: https://execution.metamask.io/iframe/6.5.0/index.html
+      - IFRAME_EXECUTION_ENVIRONMENT_URL: http://localhost:6363
```
Comment (FrederikBolding) [inline builds.yml:29]: Should be reverted before merge

File: package.json [hunk 2]
```diff
@@ -350,7 +350,7 @@
     "@metamask/rpc-errors": "^6.2.1",
     "@metamask/safe-event-emitter": "^3.1.1",
     "@metamask/scure-bip39": "^2.0.3",
-    "@metamask/selected-network-controller": "^15.0.2",
+    "@metamask/selected-network-controller": "npm:@metamask-previews/selected-network-controller@17.0.0-preview-55328a6",
```
Comment (FrederikBolding) [inline package.json:353]: Should be reverted before merge

File: app/scripts/lib/rpc-method-middleware/handlers/ethereum-chain-utils.js [hunk 3]
```diff
@@ -214,7 +216,11 @@ export async function switchChain(
         permissionedChainIds === undefined ||
         !permissionedChainIds.includes(chainId)
       ) {
-        await requestPermittedChainsPermission([chainId]);
+        if (isSnapId(origin)) {
```
Comment (FrederikBolding) [inline app/scripts/lib/rpc-method-middleware/handlers/ethereum-chain-utils.js:219]: Can we test this functionality?

File: app/scripts/metamask-controller.js [hunk 4]
```diff
@@ -5502,6 +5503,25 @@ export default class MetamaskController extends EventEmitter {
...
+              approvedChainIds: chainIds,
+            },
+            approvedPermissions: {
+              [PermissionNames.permittedChains]: {
+                caveats: [
+                  CaveatFactories[CaveatTypes.restrictNetworkSwitching](
+                    chainIds,
+                  ),
+                ],
+              },
```
Comment (FrederikBolding) [inline app/scripts/metamask-controller.js:5521]: Seems a bit strange that we have to specify the chain IDs twice 🤔 Seems like the permission specification strictly uses `requestData`

File: app/scripts/migrations/126.ts [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+import { CaveatTypes } from '../../../shared/constants/permissions';
+import { BUILT_IN_NETWORKS } from '../../../shared/constants/network';
+
+type GenericPermissionControllerSubject =
+  PermissionControllerSubjects<PermissionConstraint>[string];
+
+export const version = 126;
+
+/**
+ * This migration adds the `permittedChains` permission to all Snaps that have
```
Comment (FrederikBolding) [inline app/scripts/migrations/126.ts:20]:
```suggestion
 * This migration adds the `endowment:permittedChains` permission to all Snaps that have
```

File: app/scripts/migrations/126.ts [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+  ][];
+  permissionControllerState.subjects = entries.reduce<
+    PermissionControllerSubjects<PermissionConstraint>
+  >((accumulator, [key, subject]) => {
+    const permissionKeys = Object.keys(subject.permissions);
+    const needsMigration =
+      permissionKeys.includes(SnapEndowments.EthereumProvider) &&
+      !permissionKeys.includes(PermissionNames.permittedChains);
+    if (!needsMigration) {
+      return accumulator;
```
Comment (FrederikBolding) [inline app/scripts/migrations/126.ts:167]: Doesn't this delete any subjects that don't need the migration?

File: app/scripts/migrations/126.test.ts [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+    const oldStorage = {
+      meta: { version: oldVersion },
+      data: {},
+    };
+
+    const newStorage = await migrate(oldStorage);
+    expect(newStorage.meta).toStrictEqual({ version });
+  });
+
+  it('adds the network endowment to Snaps with the `endowment:ethereum-provider` permission', async () => {
```
Comment (FrederikBolding) [inline app/scripts/migrations/126.test.ts:64]: Test that it doesn't touch Snaps that don't have it

File: app/scripts/metamask-controller.js [hunk 8]
```diff
@@ -6204,9 +6205,11 @@ export default class MetamaskController extends EventEmitter {
         // network configuration-related
         setActiveNetwork: async (networkClientId) => {
           await this.networkController.setActiveNetwork(networkClientId);
-          // if the origin has the eth_accounts permission
-          // we set per dapp network selection state
+
+          // If the origin has the `eth_accounts` permission, or if it's a Snap,
+          // we set per origin network selection state.
           if (
+            subjectType === SubjectType.Snap ||
```
Comment (FrederikBolding) [inline app/scripts/metamask-controller.js:6212]: What is the reasoning for this?

File: app/scripts/migrations/139.ts [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    GenericPermissionControllerSubject,
+  ][];
+
+  permissionControllerState.subjects = entries.reduce<
+    PermissionControllerSubjects<PermissionConstraint>
+  >((accumulator, [key, subject]) => {
+    const permissionKeys = Object.keys(subject.permissions);
+    const needsMigration =
+      permissionKeys.includes(SnapEndowments.EthereumProvider) &&
+      !permissionKeys.includes(PermissionNames.permittedChains);
```
Comment (FrederikBolding) [inline app/scripts/migrations/139.ts:115]: AFAIK there is no way that the Snap already has `permittedChains`?

File: app/scripts/metamask-controller.js (hunk 8 above)
Comment (Mrtenz) [reply app/scripts/metamask-controller.js:6212]: `setActiveNetwork` checks if the origin has the `eth_accounts` permission to see if the network should be set for that specific origin. Snaps may not (cannot?) have this permission, but we want to use the per-origin-network regardless for Snaps.

**Comment by FrederikBolding** [comment]
@adonesky1 or @jiexi should review this PR too

File: app/scripts/migrations/139.ts [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+      };
+    }
+
+    updatedSubjects.push(key);
+
+    const newSubject: GenericPermissionControllerSubject = {
+      ...subject,
+      permissions: {
+        ...subject.permissions,
+        [PermissionNames.permittedChains]: {
```
Comment (FrederikBolding) [inline app/scripts/migrations/139.ts:130]: This permission has been removed in favor of CAIP-25, so we'll need to adjust this slightly

File: app/scripts/migrations/140.ts [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+    GenericPermissionControllerSubject,
+  ][];
+
+  permissionControllerState.subjects = entries.reduce<
+    PermissionControllerSubjects<PermissionConstraint>
+  >((accumulator, [key, subject]) => {
+    const permissionKeys = Object.keys(subject.permissions);
+    const needsMigration =
+      permissionKeys.includes(SnapEndowments.EthereumProvider) &&
+      !permissionKeys.includes(PermissionNames.permittedChains);
```
Comment (adonesky1) [inline app/scripts/migrations/140.ts:167]: This looks to be out of date since [this PR was merged yesterday](https://github.com/MetaMask/metamask-extension/pull/27847). `PermissionNames.permittedChains` no longer exists. Instead for this check we will need to examine whether it has a `Caip25EndowmentPermissionName`.

File: app/scripts/migrations/140.ts [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+    updatedSubjects.push(key);
+
+    const caveatValue = addPermittedEthChainId(
+      {
+        requiredScopes: {},
+        optionalScopes: {},
+        isMultichainOrigin: false,
+      },
+      currentChainId as `0x${string}`,
+    );
```
Comment (adonesky1) [inline app/scripts/migrations/140.ts:185]: I think this logic will have to account for whether or not there are also existing account permissions, because this will overwrite those permissions if so.

Comment (adonesky1) [reply app/scripts/migrations/140.ts:185]: These account permissions will have been migrated into the CAIP-25 permission already: https://github.com/MetaMask/metamask-extension/pull/27847/files#diff-e63a7581f2f988fe114f58ecfaa796e1f6efc4f8934e5883dfca9204618edb13R402

File: app/scripts/migrations/140.ts [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+ * @returns The updated MetaMask extension state.
+ */
+function transformState(state: Record<string, unknown>) {
+  const {
+    NetworkController: networkControllerState,
+    PermissionController: permissionControllerState,
+    SelectedNetworkController: selectedNetworkControllerState,
+  } = state;
+  if (!networkControllerState || !isObject(networkControllerState)) {
+    log.warn(
```
Comment (adonesky1) [inline app/scripts/migrations/140.ts:84]: 🤔 where do these logs go? I'm more familiar with the pattern of sending these logs to sentry via:
`global.sentry?.captureException(...)`. Like [here for example](https://github.com/MetaMask/metamask-extension/blob/f856a3b423a4316f14d2e5d7cd0297f76566f70a/app/scripts/migrations/139.ts#L168).

File: app/scripts/migrations/140.ts [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+    string,
+    GenericPermissionControllerSubject,
+  ][];
+
+  permissionControllerState.subjects = entries.reduce<
+    PermissionControllerSubjects<PermissionConstraint>
+  >((accumulator, [key, subject]) => {
+    const permissionKeys = Object.keys(subject.permissions);
+    const needsMigration =
+      permissionKeys.includes(SnapEndowments.EthereumProvider) &&
```
Comment (adonesky1) [inline app/scripts/migrations/140.ts:166]: I'm assuming this permission can only be granted to Snaps? Otherwise we could be overwriting permissions for non-snaps subjects?

File: app/scripts/lib/rpc-method-middleware/handlers/ethereum-chain-utils.js [hunk 15]
```diff
@@ -201,13 +204,13 @@ export async function switchChain(
       if (!ethChainIds.includes(chainId)) {
         await requestPermittedChainsPermissionIncrementalForOrigin({
           chainId,
-          autoApprove: isAddFlow,
+          autoApprove: isAddFlow || isSnapId(origin),
```
Comment (jiexi) [inline app/scripts/lib/rpc-method-middleware/handlers/ethereum-chain-utils.js:207]: i have a PR here that replaces the `isAddFlow` param with one named `autogrant`. Perhaps this isSnapId check can happen outside the `switchChain` helper that way we don't need to pass `origin` in?

File: app/scripts/migrations/140.ts [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+    return state;
+  }
+
+  const currentChainId = getChainId(
+    networkControllerState.selectedNetworkClientId,
+    networkControllerState.networkConfigurationsByChainId,
+  );
+  if (typeof currentChainId !== 'string') {
+    log.warn(
+      'Skipping migration: `NetworkController.selectedNetworkClientId` is not a string.',
```
Comment (jiexi) [inline app/scripts/migrations/140.ts:147]:
```suggestion
      'Skipping migration: The chainId resolved from the `NetworkController.selectedNetworkClientId` value is not a string.',
```

Comment (jiexi) [reply app/scripts/migrations/140.ts:147]: or something different anyway

File: app/scripts/lib/rpc-method-middleware/handlers/ethereum-chain-utils.js (hunk 15 above)
Comment (Mrtenz) [reply app/scripts/lib/rpc-method-middleware/handlers/ethereum-chain-utils.js:207]: Makes sense. I didn't want to use `isAddFlow` since it's not an add flow, but `autoApprove` sounds good.

File: app/scripts/migrations/140.ts (hunk 13 above)
Comment (Mrtenz) [reply app/scripts/migrations/140.ts:84]: Not sure. I copied this from another migration, but I see now that the Sentry pattern is used in others. I'll swap it out for that.

File: app/scripts/migrations/140.ts (hunk 11 above)
Comment (Mrtenz) [reply app/scripts/migrations/140.ts:167]: Good catch, I forgot to update the permission name here.

File: app/scripts/migrations/140.ts (hunk 14 above)
Comment (Mrtenz) [reply app/scripts/migrations/140.ts:166]: Yes, it can only be granted to Snaps.

File: app/scripts/lib/rpc-method-middleware/handlers/ethereum-chain-utils.js (hunk 15 above)
Comment (jiexi) [reply app/scripts/lib/rpc-method-middleware/handlers/ethereum-chain-utils.js:207]: i managed to not link the PR lol. Here it is

https://github.com/MetaMask/metamask-extension/pull/29837

File: app/scripts/migrations/144.ts [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+ */
+const BUILT_IN_NETWORKS: ReadonlyMap<string, Hex> = new Map([
+  ['sepolia', '0xaa36a7'],
+  ['mainnet', '0x1'],
+  ['linea-sepolia', '0xe705'],
+  ['linea-mainnet', '0xe708'],
+]);
+
+/**
+ * This migration adds the `endowment:permitted-chains` permission to all Snaps
```
Comment (FrederikBolding) [inline app/scripts/migrations/144.ts:35]:
```suggestion
 * This migration adds the `endowment:caip25` permission to all Snaps
```

File: app/scripts/migrations/144.ts [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+  ['linea-mainnet', '0xe708'],
+]);
+
+/**
+ * This migration adds the `endowment:caip25` permission to all Snaps
+ * that have the `endowment:ethereum-provider` permission, and sets the selected
+ * chain ID for each Snap to the current selected network chain ID.
+ *
+ * This is necessary following the Amon Hen v2 changes to the network controller
+ * and other related controllers, as the `endowment:permitted-chains` permission
```
Comment (FrederikBolding) [inline app/scripts/migrations/144.ts:40]:
```suggestion
 * and other related controllers, as the `endowment:caip25` permission
```

These comments may need more updating though, since the CAIP25 permission is already added, we just wanna add the chain ID

File: app/scripts/migrations/144.ts [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+ * is now required to use the Ethereum provider and switch networks.
+ *
+ * To simplify the use for Snaps, we automatically add the
+ * `endowment:permitted-chains` permission with the current selected network
```
Comment (FrederikBolding) [inline app/scripts/migrations/144.ts:44]:
```suggestion
 * `endowment:caip25` permission with the current selected network
```

File: app/scripts/migrations/144.ts [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+
+  if (!existingCaveat) {
+    return {
+      isMultichainOrigin: false,
+      optionalScopes: {},
+      requiredScopes: {},
+    };
+  }
+
+  return existingCaveat.value as Caip25CaveatValue;
```
Comment (adonesky1) [inline app/scripts/migrations/144.ts:272]: Could we remove this type cast?

File: app/scripts/migrations/144.ts [hunk 21]
```diff
@@ -0,0 +1,301 @@
...
+
+  const currentChainId = getChainId(
+    networkControllerState.selectedNetworkClientId,
+    networkControllerState.networkConfigurationsByChainId,
+  );
+
+  if (typeof currentChainId !== 'string') {
+    global.sentry?.captureException?.(
+      new Error(
+        'Skipping migration: The chain ID resolved from the `NetworkController.selectedNetworkClientId` value is not a string.',
```
Comment (adonesky1) [inline app/scripts/migrations/144.ts:171]: I feel like this error message is slightly misleading since it will also throw when no chainId is found?

File: app/scripts/migrations/144.ts [hunk 22]
```diff
@@ -0,0 +1,301 @@
...
+        ...accumulator,
+        [key]: subject,
+      };
+    }
+
+    updatedSubjects.push(key);
+
+    const caveatValue = addPermittedEthChainId(
+      getCaveat(subject),
+      currentChainId as `0x${string}`,
```
Comment (adonesky1) [inline app/scripts/migrations/144.ts:206]: We should be able to get rid of this typecast, `currentChainId` should be of type `Hex`

File: app/scripts/migrations/144.ts [hunk 23]
```diff
@@ -0,0 +1,301 @@
...
+          caveats: [
+            {
+              type: Caip25CaveatType,
+              value: caveatValue,
+            },
+          ],
+          date: Date.now(),
+          id: nanoid(),
+          invoker: key,
+          parentCapability: Caip25EndowmentPermissionName,
```
Comment (adonesky1) [inline app/scripts/migrations/144.ts:223]: probably doesn't matter but should we inherit this metadata from the existing CAIP-25 permission if it exists?

File: app/scripts/migrations/144.ts (hunk 20 above)
Comment (Mrtenz) [reply app/scripts/migrations/144.ts]: If there's a type guard we could remove it. The caveat is a `CaveatConstraint`, TypeScript can't infer that it's a CAIP-25 permission here.

File: app/scripts/migrations/151.ts [hunk 24]
```diff
@@ -0,0 +1,308 @@
...
+ * Get the existing `endowment:caip25` caveat value for a given subject, if any.
+ * If the caveat does not exist, a new empty caveat value is returned.
+ *
+ * This function does not validate the structure of the existing caveat value,
+ * and assumes that the existing value is valid.
+ *
+ * @param subject - The subject to get the existing caveat value for.
+ * @returns The existing `endowment:caip25` caveat value.
+ */
+function getCaveat(
```
Comment (adonesky1) [inline app/scripts/migrations/151.ts:264]: maybe?
```suggestion
function getExistingCaip25PermissionCaveat(
```

File: app/scripts/migrations/151.ts [hunk 25]
```diff
@@ -0,0 +1,308 @@
...
+    if (!needsMigration) {
+      return {
+        ...accumulator,
+        [key]: subject,
+      };
+    }
+
+    updatedSubjects.push(key);
+
+    const caveatValue = addPermittedEthChainId(
```
Comment (adonesky1) [inline app/scripts/migrations/151.ts:204]: maybe
```suggestion
    const caip25PermissionCaveatWithCurrentChainIdSet = addPermittedEthChainId(
```

Comment (adonesky1) [reply app/scripts/migrations/151.ts:204]: I realize these var names don't matter that much since this is a migration, but helps me keep track better of what we're referencing

File: app/scripts/migrations/151.ts [hunk 26]
```diff
@@ -0,0 +1,308 @@
...
+
+  if (!existingCaveat) {
+    return {
+      isMultichainOrigin: false,
+      optionalScopes: {},
+      requiredScopes: {},
+    };
+  }
+
+  return existingCaveat.value as Caip25CaveatValue;
```
Comment (adonesky1) [inline app/scripts/migrations/151.ts:279]: Just curious how could this ever be hit?

Comment (adonesky1) [reply app/scripts/migrations/151.ts:279]: And if it is hit should we check what chainId/scopes are already set in the caveat?

Comment (Mrtenz) [reply app/scripts/migrations/151.ts:279]: From my understanding `eth_accounts` uses the same caveat, does it not?

File: app/scripts/migrations/151.ts [hunk 27]
```diff
@@ -0,0 +1,308 @@
...
+}
+
+function getChainId(
+  networkClientId: string,
+  networkConfigurationsByChainId: RuntimeObject,
+): Hex | undefined {
+  const configurationChainId = Object.entries(
```
Comment (adonesky1) [inline app/scripts/migrations/151.ts:286]:
```suggestion
  const matchingNetworkConfigurationEntry = Object.entries(
```

File: app/scripts/migrations/151.ts [hunk 28]
```diff
@@ -0,0 +1,308 @@
...
+
+    return config.rpcEndpoints.some(
+      (endpoint) =>
+        isObject(endpoint) && endpoint.networkClientId === networkClientId,
+    );
+  });
+
+  if (configurationChainId) {
+    return configurationChainId[0] as Hex;
+  }
```
Comment (adonesky1) [inline app/scripts/migrations/151.ts]:
```suggestion
  if (matchingNetworkConfigurationEntry) {
  // [0] in the entry is the chainId key
    return matchingNetworkConfigurationEntry[0] as Hex;
  }
```

File: app/scripts/migrations/151.ts (hunk 26 above)
Comment (adonesky1) [reply app/scripts/migrations/151.ts:279]: yes it does now...  I'm forgetting though have we already granted snaps the caip25 permission elsewhere?

Comment (Mrtenz) [reply app/scripts/migrations/151.ts:279]: Not to my knowledge, I think we only use it for accounts.

Comment (adonesky1) [reply app/scripts/migrations/151.ts:279]: I thought this was the first instance of us giving snaps this new permission?

Comment (adonesky1) [reply app/scripts/migrations/151.ts:279]: sorry nvm

File: app/scripts/migrations/151.test.ts [hunk 29]
```diff
@@ -0,0 +1,303 @@
...
+                    {
+                      type: Caip25CaveatType,
+                      value: {
+                        isMultichainOrigin: false,
+                        optionalScopes: {
+                          'eip155:59141': {
+                            accounts: [
+                              '0x1234567890123456789012345678901234567890',
+                            ],
+                          },
```
Comment (adonesky1) [inline app/scripts/migrations/151.test.ts:170]: Just to be realistic, prior to this migration the only scope that this caveat would have is `wallet:eip155` and no accounts
```suggestion
                          'waillet:eip155': {
                            accounts: [],
                          },
```

File: app/scripts/migrations/151.test.ts [hunk 30]
```diff
@@ -0,0 +1,303 @@
...
+                  value: {
+                    isMultichainOrigin: false,
+                    optionalScopes: {
+                      'eip155:1': {
+                        accounts: [],
+                      },
+                      'eip155:59141': {
+                        accounts: [
+                          '0x1234567890123456789012345678901234567890',
+                        ],
```
Comment (adonesky1) [inline app/scripts/migrations/151.test.ts:219]:
```suggestion
                      'wallet:eip155': {
                        accounts: [],
```

**Review by adonesky1 — DISMISSED** [review]
LGTM!

File: app/scripts/migrations/151.test.ts [hunk 31]
```diff
@@ -0,0 +1,303 @@
...
+            },
+          },
+        },
+      },
+    });
+
+    expect(newStorage.data.SelectedNetworkController).toStrictEqual({
+      domains: {
+        [METAMASK_DOMAIN]: 'mainnet',
+        [MOCK_SNAP_ID]: '0xe705',
```
Comment (jiexi) [inline app/scripts/migrations/151.test.ts:119]: shouldn't this match the networkClientId, not the chainId?

File: app/scripts/migrations/151.ts [hunk 32]
```diff
@@ -0,0 +1,310 @@
...
+
+    return {
+      ...accumulator,
+      [key]: newSubject,
+    };
+  }, {});
+
+  const currentDomains = selectedNetworkControllerState.domains;
+  const domains = Object.fromEntries(
+    updatedSubjects.map((subject) => [subject, currentChainId]),
```
Comment (jiexi) [inline app/scripts/migrations/151.ts:243]: pretty sure the value of this should be a networkClientId, not chainId

Comment (adonesky1) [reply app/scripts/migrations/151.ts:243]: oooof good catch Jiexi! yeah I think this should be something like:
```suggestion
    updatedSubjects.map((subject) => [subject, networkControllerState.selectedNetworkClientId]),
```

Comment (Mrtenz) [reply app/scripts/migrations/151.ts:243]: I thought I checked that both work, but maybe not. Will update! The network controller and selected network controller can be hard to follow, especially in migrations where we can't really rely on the controllers directly 😅

**Comment by adonesky1** [comment]
Could you explain a bit more directly in the PR description how to:
> Use wallet_switchEthereumChain in a Snap.
?

**Comment by Mrtenz** [comment]
> Could you explain a bit more directly in the PR description how to:
> 
> > Use wallet_switchEthereumChain in a Snap.
> 
> ?

@adonesky1 I've updated the instructions to be more detailed!

File: app/scripts/metamask-controller.test.js [hunk 33]
```diff
@@ -1243,90 +1243,6 @@ describe('MetaMaskController', () => {
         );
       });
 
-      it('returns approval from the PermissionsController for only eth_accounts when only permittedChains is specified in params and origin is snapId', async () => {
```
Comment (adonesky1) [inline app/scripts/metamask-controller.test.js:1246]: should we preserve these tests but redo them as 
`returns approval from the PermissionsController for both eth_accounts and permittedChains when only permittedChains is specified in params and origin is snapId` ?

Comment (adonesky1) [reply app/scripts/metamask-controller.test.js:1246]: So covering the snap case with the new functionality?

File: app/scripts/migrations/152.ts [hunk 34]
```diff
@@ -0,0 +1,313 @@
...
+      requiredScopes: {},
+    };
+  }
+
+  return existingCaveat.value as Caip25CaveatValue;
+}
+
+function getChainId(
+  networkClientId: string,
+  networkConfigurationsByChainId: RuntimeObject,
```
Comment (adonesky1) [inline app/scripts/migrations/152.ts:287]: Any reason we can't type this as:
`networkConfigurationsByChainId: Record<string, NetworkConfiguration>` ?

**Review by FrederikBolding — APPROVED** [review]
